### PR TITLE
Fix concurrent checks limit while using command_endpoint

### PR DESCRIPTION
**Pull request by N-o-X**

### CODE REVIEW

File: doc/09-object-types.md [hunk 1]
```diff
@@ -249,16 +249,14 @@ Example:
...
-}
+object CheckerComponent "checker" { }
 ```
 
 Configuration Attributes:
 
   Name                      | Type                  | Description
   --------------------------|-----------------------|----------------------------------
-  concurrent\_checks        | Number                | **Optional.** The maximum number of concurrent checks. Defaults to 512.
+  concurrent\_checks        | Number                | **Optional and Deprecated.** The maximum number of concurrent checks. Was geplaced by global constant 'MaxConcurrentChecks' which will be set if you still use 'concurrent_checks'.
```
Comment (dnsmichi) [inline doc/09-object-types.md]: `geplaced` typo.

File: lib/icinga/clusterevents-check.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	m_CheckSchedulerRunning = false;
+}
+
+void ClusterEvents::EnqueueCheck(const MessageOrigin::Ptr& origin, const Dictionary::Ptr& params)
+{
+	boost::mutex::scoped_lock lock(m_Mutex);
+
+	if (m_CheckRequestQueue.size() >= 25000) {
```
Comment (dnsmichi) [inline lib/icinga/clusterevents-check.cpp]: Where does this value come from?

**Comment by dnsmichi** [comment]
Short notes:

- v2.9 only, no backports
- this needs reliable long-term tests
- more metrics and insights (e.g. if the remove queue limit is hit, can I have that via /v1/status somehow?)

File: doc/09-object-types.md [hunk 3]
```diff
@@ -249,16 +249,14 @@ Example:
...
-}
+object CheckerComponent "checker" { }
 ```
 
 Configuration Attributes:
 
   Name                      | Type                  | Description
   --------------------------|-----------------------|----------------------------------
-  concurrent\_checks        | Number                | **Optional.** The maximum number of concurrent checks. Defaults to 512.
+  concurrent\_checks        | Number                | **Optional and Deprecated.** The maximum number of concurrent checks. Was replaced by global constant 'MaxConcurrentChecks' which will be set if you still use 'concurrent_checks'.
```
Comment (gunnarbeutner) [inline doc/09-object-types.md]: Those should probably be backticks.

File: lib/icinga/clusterevents-check.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+#include "base/exception.hpp"
+#include <boost/thread/once.hpp>
+
+using namespace icinga;
+
+boost::mutex ClusterEvents::m_Mutex;
+std::deque<std::function<void ()>> ClusterEvents::m_CheckRequestQueue;
+bool ClusterEvents::m_CheckSchedulerRunning;
+int ClusterEvents::m_ChecksExecutedDuringInterval;
+static Timer::Ptr logTimer;
```
Comment (gunnarbeutner) [inline lib/icinga/clusterevents-check.cpp]: Can you rename this to l_CheckRequestQueueTimer or something like that?

File: lib/icinga/clusterevents.hpp [hunk 5]
```diff
@@ -25,6 +25,7 @@
 #include "icinga/checkcommand.hpp"
 #include "icinga/eventcommand.hpp"
 #include "icinga/notificationcommand.hpp"
+#include <thread>
```
Comment (Crunsher) [inline lib/icinga/clusterevents.hpp]: Please move this to the cpp

**Comment by dnsmichi** [comment]
I'll put this on my test list for tomorrow.

File: lib/icinga/cib.cpp [hunk 6]
```diff
@@ -305,6 +305,8 @@ void CIB::StatsFunc(const Dictionary::Ptr& status, const Array::Ptr& perfdata) {
 	status->Set("active_service_checks_15min", GetActiveServiceChecksStatistics(60 * 15));
 	status->Set("passive_service_checks_15min", GetPassiveServiceChecksStatistics(60 * 15));
 
+	status->Set("remote_check_queue", ClusterEvents::GetCheckRequestQueueSize());
```
Comment (dnsmichi) [inline lib/icinga/cib.cpp]: This breaks non-unity builds as the header include is missing.

**Comment by dnsmichi** [comment]
# Status

@lippserd 

TODO

- [ ] docs
- [ ] logging when the queue is full
- [ ] long term performance tests 

# Docs

Documentation inside the Distributed Monitoring chapter is entirely missing. I had to figure it out myself from reading the code. We should fix this prior to merging.

# Logging

Once the remote execution queue is full, this spams the log every time. We should carefully handle this. Needs to be fixed before this gets merged.

screen shot 2018-01-19 at 17 19 44


```
[2018-01-19 17:17:03 +0100] information/ApiListener: Finished reconnecting to endpoint 'icinga2b' via host 'localhost' and port '8000'
[2018-01-19 17:17:03 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:17:03 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:17:03 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:17:03 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.

<insert 1000 lines here>

[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:04 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:10 +0100] information/RemoteCheckQueue: items: 25000, rate: 51/s (3072/min 15360/5min 46080/15min);
[2018-01-19 17:18:11 +0100] critical/ClusterEvents: Remote check queue ran out of slots. Discarding remote check request.
[2018-01-19 17:18:13 +0100] information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 1/15min);
[2018-01-19 17:18:13 +0100] information/WorkQueue: #5 (ApiListener, RelayQueue) items: 0, rate: 0.566667/s (34/min 83/5min 122/15min);
[2018-01-19 17:18:13 +0100] information/WorkQueue: #11 (JsonRpcConnection, #2) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2018-01-19 17:18:13 +0100] information/WorkQueue: #10 (JsonRpcConnection, #1) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2018-01-19 17:18:13 +0100] information/WorkQueue: #12 (JsonRpcConnection, #3) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2018-01-19 17:18:13 +0100] information/WorkQueue: #9 (JsonRpcConnection, #0) items: 0, rate: 51.5167/s (3091/min 33763/5min 33787/15min);
```

# Test Config

Use many.conf but execute long lasting checks (sleep for a while).

On the client, set `const MaxConcurrentChecks = 512` into constants.conf or likewise.

```
mbmif /usr/local/tests/icinga2/master-slave (master *+) # cat icinga2c.conf | grep Conc
const MaxConcurrentChecks = 512
```

```
mbmif /usr/local/tests/icinga2/master-slave (master *+) # cat icinga2a/etc/icinga2/zones.d/global-templates/commands.conf
object CheckCommand "sleep" {
  command = [ "/bin/sleep", 30 ]
}
```

```
mbmif /usr/local/tests/icinga2/master-slave (master *+) # cat icinga2a/etc/icinga2/zones.d/master/many.conf
const countHosts = 10;
const countServices = 1000;


for (i in range(countHosts)) {
  object Host "many-test-" + i {
    check_command = "random"
  }
}

for (j in range(countServices)) {
  apply Service "many-test-" + j {
    //check_command = "random"
    check_command = "sleep"
    command_endpoint = "icinga2c" //command endpoint test
    check_interval = 1s
    retry_interval = 1s
    assign where match("many*", host.name)
  }
}
```

**Comment by N-o-X** [comment]
Added documentation for the remote check queue and fixed logging.

**Comment by dnsmichi** [comment]
screen shot 2018-02-02 at 16 22 18

File: doc/09-object-types.md [hunk 7]
```diff
@@ -247,16 +247,16 @@ This configuration object is available as [checker feature](11-cli-commands.md#c
 Example:
 
 ```
-object CheckerComponent "checker" {
-  concurrent_checks = 512
-}
+library "checker"
```
Comment (dnsmichi) [inline doc/09-object-types.md]: That's obsolete with v2.9.

File: doc/06-distributed-monitoring.md [hunk 8]
```diff
@@ -818,6 +818,10 @@ This mode will force the Icinga 2 node to execute commands remotely on a specifi
 The host/service object configuration is located on the master/satellite and the client only
 needs the CheckCommand object definitions being used there.
 
+Every endpoint has its own remote check queue. The amount of checks executed simultaneously
+is limited on the endpoint by `MaxConcurrentChecks`. Icinga 2 may discard check requests,
```
Comment (dnsmichi) [inline doc/06-distributed-monitoring.md]: .... can be limited with the `MaxConcurrentChecks` constant defined in [constants.conf](#constants-conf) 

Or similar.

**Pull request merged**